clarity-ext-scripts/clarity_ext_scripts/covid/rtpcr_analyse_script.py

```diff
@@ -1,60 +1,64 @@
-RT_PCR_POSITIVE_CONTROL = 'rtpcr_pos'
-RT_PCR_NEGATIVE_CONTROL = 'rtpcr_neg'
+
+RT_PCR_POSITIVE_CONTROL = 'Positive PCR Control'
+RT_PCR_NEGATIVE_CONTROL = 'Negative PCR Control'
 
 
 class Extension(GeneralExtension):
     def execute(self):
-        self.initial_validation()
-        ct_analysis_service = RTPCRAnalysisService()
-        lower_bound, upper_bound = BOUNDARY_DICT[self.assay]
-
-        rt_pcr_control_types = [RT_PCR_POSITIVE_CONTROL, RT_PCR_NEGATIVE_CONTROL]
-        found_controls = list()
-        self.context.current_step.udf_map.force("rtPCR Passed", False)
-        for _, output in self.context.all_analytes:
-            original_sample = output.sample()
-            original_sample.udf_map.force("rtPCR Passed", False)
+        if not self._has_assay_udf():
+            raise UsageError("The udf 'Assay' must be filled in before running this script")
 
-        # 1. Check control values
+        # Prepare analyse service input args
+        samples = list()
+        positive_controls = list()
+        negative_controls = list()
         for _, output in self.context.all_analytes:
-            original_sample = output.sample()
-            if original_sample.udf_control == 'Yes' \
-                    and original_sample.udf_control_type.lower() in rt_pcr_control_types:
-                found_controls.append(original_sample.udf_control_type.lower())
-                ct_analysis_service.validate_control_value(
-                    original_sample.udf_control_type, output.udf_ct,
-                    lower_bound, upper_bound)
+            # TODO: currently we only have udf for CT (no distinction between FAM and HEX)
+            result = CTResult(id=output.id, fam_ct=output.udf_famct, hex_ct=output.udf_hexct)
+            if output.name == RT_PCR_POSITIVE_CONTROL:
+                positive_controls.append(result.get_dict())
+            elif output.name == RT_PCR_NEGATIVE_CONTROL:
+                negative_controls.append(result.get_dict())
+            else:
+                samples.append(result.get_dict())
 
-        if not set(rt_pcr_control_types).issubset(set(found_controls)):
-            raise UsageError('positive and negative rtPCR controls were not found on this plate: {}'
-                             .format(set(found_controls)))
+        if len(positive_controls) == 0 or len(negative_controls) == 0:
+            raise UsageError('positive and negative rtPCR controls were not found on this plate.')
 
-        if not ct_analysis_service.is_valid():
-            return
+        # Check control values
+        ct_analysis_service = RTPCRAnalysisService()
+        self.context.current_step.udf_map.force("rtPCR Passed", True)
+        self.context.update(self.context.current_step)
+        try:
+            ct_analysis_service._analyze_controls(positive_controls, negative_controls)
+        except MultipleAnalysisErrors:
+            self.context.current_step.udf_map.force("rtPCR Passed", False)
```

Comment:
Why run the analysis part twice? I think it should be designed in such a way that you wan put a `try...except` block around the whole thing and use that to handle both the "happy" and the "failure" part?

Reply:
Looking back at the old code I understand better why you needed to do it this way... Though I think it would be a good idea to simplify this now that the service it better behaved (it doesn't raise errors in quite the same what it did before).

Reply:
You can regard this PR as WIP. I realized that the service wasn't finished all the way, so I tried to do the best from the interface that was at hand.